lib/report_generator.rb

```diff
@@ -60,23 +57,24 @@ def build_report_pms(query, table = "", usernames = [])
       pms
     end
 
-    private
-
-    def filter_recipients_by_query_access(recipients, query)
-      recipients.reduce([]) do |names, recipient|
-        if (group = Group.find_by(name: recipient)) &&
-             (
-               group.id == Group::AUTO_GROUPS[:admins] ||
-                 query.query_groups.exists?(group_id: group.id)
-             )
-          names.concat group.users.pluck(:username)
-        elsif (user = User.find_by(username: recipient)) &&
-              Guardian.new(user).user_can_access_query?(query)
-          names << recipient
+    def self.filter_recipients_by_query_access(recipients, query)
+      recipients
+        .uniq
+        .reduce([]) do |names, recipient|
+          if (group = Group.find_by(name: recipient)) &&
```

Comment:
This one and others (like User.find_by) feels like an N+1 no? (I see it was already present before)

Reply:
My expectation is that there will likely only be 3-4 recipients entered, and the script is only used by site admins. But I'm open to refactoring this if it's a concern that it could be misused.

Multiple recipient types are here so an alternative might be to do `.where` against groups and users, then loop the results against Guardian and remove those that fail.

Reply:
Yes if you can try to optimize this a little bit that would be great please